scripts: add support for IRONside SE update service (#22065)
nrfconnect/sdk-nrf: applications/ironside/se/update/README.rst

.. _ironside_se_update_application:

> IRONside SE Update application
> ##############################

Reviewer: ```suggestion
IRONside SE Update
##################
```

> This application is used to trigger an update of the IRONside SE firmware.
> The new firmware can be found in the release zip of the IRONside SE.
> Both recovery and main IRONside SE firmwares can be updated.

Reviewer: ```suggestion
The *IRONside SE Update* application updates the IRONside SE firmware on a SoC that already has IRONside SE installed.
It can update both the main image and the recovery image of IRONside SE using the IRONside SE firmware release ZIP file.
```

Update procedure
****************

> IRONside SE firmware update hex files are located at a known address.
> This application invokes the IRONside SE update service, passing the parameters corresponding to the update hex file.
> The return value of the service call is printed, as well as information from the update hex file.
> Once the service has been invoked, the IRONside SE firmware has updated the internal state of the device so that the update will be installed on the next boot.

Reviewer: ```suggestion
The application invokes the IRONside SE update service and passes the parameters that correspond to the location of the HEX file of the IRONside SE firmware update.
   
The application prints the return value of the service call and outputs information from the update HEX file.

After the service call completes, the IRONside SE firmware updates the internal state of the device.
The firmware installs the update during the next device boot.
```

> The firmware installs the update during the next device boot.

Reviewer 2: I'd add that the update process needs a short time period and can be verified with reading back the version via the boot report

Requirements
************

The firmware supports the following development kit:

.. table-from-sample-yaml::

Building and running
********************

> This application can only be used when there is already a version of IRONside SE installed on the device.

Reviewer: ```suggestion
.. note::
   You can use this application only when a version of IRONside SE is already installed on the device.
```

Reviewer: @hakonfam Can you build and program this app using VScode?

Author: The IRONside SE app? No.

Reviewer: Please add this note

> Start with unzipping the release zip to get the update hex file.
> 
>   .. code-block:: console
> 
>      unzip nrf54h20_soc_binaries_v.*.zip
> 
> Build and program this application
> 
>   .. code-block:: console
> 
>      west build -b nrf54h20dk/nrf54h20/cpuapp/iron
>      west flash
> 
> Program the appropriate update hex file:
> 
>   .. code-block:: console
> 
>      # Perform one of the following commands
>      nrfutil device program --traits jlink --firmware update/ironside_se_update.hex # to update IRONside SE firmware
>      nrfutil device program --traits jlink --firmware update/ironside_se_recovery_update.hex # to update IRONside SE recovery firmware
> 
> Trigger a reset
> 
>   .. code-block:: console
> 
>      # Perform one of the following commands
>      nrfutil device reset --reset-kind RESET_PIN --traits jlink

Reviewer: ```suggestion
To build and run the application, do the following:

1. Unzip the update release ZIP to get the update HEX file.

   .. code-block:: console
      unzip nrf54h20_soc_binaries_v.*.zip
     
#. Build and program this application

   .. code-block:: console
      west build -b nrf54h20dk/nrf54h20/cpuapp/iron
      west flash
     
#. Program the appropriate update hex file:

   .. code-block:: console
      # Perform one of the following commands
      nrfutil device program --traits jlink --firmware update/ironside_se_update.hex # to update IRONside SE firmware
      nrfutil device program --traits jlink --firmware update/ironside_se_recovery_update.hex # to update IRONside SE recovery firmware
     
#. Trigger a reset

   .. code-block:: console
      # Perform one of the following commands
      nrfutil device reset --reset-kind RESET_PIN --traits jlink

```

> 

Reviewer 2: potentially add steps to wait + verify via boot-report version that the update worked

.. _ironside_se_update_application:

IRONside SE Update
##################

The *IRONside SE Update* application updates the IRONside SE firmware on a SoC that already has IRONside SE installed.
It can update both the main image and the recovery image of IRONside SE using the IRONside SE firmware release ZIP file.

Update procedure
****************

The application invokes the IRONside SE update service and passes the parameters that correspond to the location of the HEX file of the IRONside SE firmware update.

The application prints the return value of the service call and outputs information from the update HEX file.

After the service call completes, the IRONside SE firmware updates the internal state of the device.
The firmware installs the update during the next device boot.
This operation can take several seconds.

> Once the operation has completed the boot report can be read to verify that the update has taken place.

Reviewer: @hakonfam who can read the boot report? and how?

********************

.. note::
   This application can only be used when there is already a version of IRONside SE installed on the device.

To build and run application, do the following:

1. Unzip the update release ZIP to get the update HEX file.

   .. code-block:: console
      unzip nrf54h20_soc_binaries_v.*.zip

#. Build and program this application

   .. code-block:: console
      west build -b nrf54h20dk/nrf54h20/cpuapp/iron
      west flash

#. Program the appropriate update hex file:

   .. code-block:: console
      # Perform one of the following commands
      nrfutil device program --traits jlink --firmware update/ironside_se_update.hex # to update IRONside SE firmware
      nrfutil device program --traits jlink --firmware update/ironside_se_recovery_update.hex # to update IRONside SE recovery firmware

#. Trigger a reset

   .. code-block:: console
      nrfutil device reset --reset-kind RESET_PIN --traits jlink

#. Check that the new version is installed

   .. code-block:: console
      # Read the version fields from the boot report
      nrfutil x-read --direct --address 0x2f88fd04 --bytes 16 --traits jlink
      # Read the update status in the boot report

>       nrfutil x-read --direct --address 0x2f88fd24 --bytes 4 --traits jlink

Reviewer 3: What do you think @hakonfam? Would you like to have a Testing section?

.. _ironside_se_update_application:

IRONside SE Update
##################

The *IRONside SE Update* application updates the IRONside SE firmware on a SoC that already has IRONside SE installed.
It can update both the main image and the recovery image of IRONside SE using the IRONside SE firmware release ZIP file.

Update procedure
****************

The application invokes the IRONside SE update service and passes the parameters that correspond to the location of the HEX file of the IRONside SE firmware update.

The application prints the return value of the service call and outputs information from the update HEX file.

After the service call completes, the IRONside SE firmware updates the internal state of the device.
The firmware installs the update during the next device boot.
This operation can take several seconds.

Once the operation has completed the boot report can be read to verify that the update has taken place.

Requirements
************

The firmware supports the following development kit:

.. table-from-sample-yaml::

Building and running
********************

This application can only be used when there is already a version of IRONside SE installed on the device.

> A version of this application is bundled with the IRONside SE release.

Reviewer: So which versions should they use? The one bundled with the ironside release or this one? @hakonfam

Author: They can choose, but I think its valuable to describe this here in case they stumble upon this sample in an attempt to perform the update.

Author: I removed the description of the script.

This can be executed by using ``west ncs-ironsiode-se-update``.

> This command will do all steps required to install the IRONside SE (or recovery) version in the zip file.

Reviewer: To install or to update? @hakonfam

Author: According to my head it "installs the update" is that valid?

Author: I removed the description of the script.

   .. code-block:: console
      # To update IRONside SE:
      west ncs-ironside-se-update --zip nrf54h20_soc_binaries_v*.zip --serial #SERIAL# --allow-erase
      # To update IRONside SE recovery:
      west ncs-ironside-se-update --zip nrf54h20_soc_binaries_v*.zip --serial #SERIAL# --allow-erase --recovery

> Alternatively, the same procedure can be performed with this application manually.

Author: The one they want, should I refrain from documenting the west command here?

> Alternatively, the same procedure can be performed with this application manually.

Author: I removed the description of the script.
